### PR TITLE
python3Packages.packageurl: init at 0.9.4

### PR DESCRIPTION
**Pull request by armijnhemel**
###### Motivation for this change

packageurl (AKA purl) is an upcoming standard to describe packages in a more uniform way across operating systems and Linux distributions

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [X] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by armijnhemel** [comment]
First Python package that I created, so some help reviewing would be much appreciated.

**Comment by pombredanne** [comment]
@armijnhemel that's awesome! I am the upstream package maintainer and I am on hand for help as needed.

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 113597` at 8f7ff002 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li><a href="https://gist.github.com/f51a0e3fbd03ed72aaea16762eda8cf7">python37Packages.packageurl-python</a></li></li>
    <li><a href="https://gist.github.com/787a0c266a8c8081207af0cd175fc250">python38Packages.packageurl-python</a></li></li>
    <li><a href="https://gist.github.com/2bdc569ef35188761dce210ef1b5f787">python39Packages.packageurl-python</a></li></li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/maintainers-missing.md">maintainers-missing</a>

Package does not have a maintainer. Consider adding yourself?

Near pkgs/development/python-modules/packageurl-python/default.nix:12:3:
```
   |
12 |   meta = with lib; {
   |   ^
```
</li>
  </ul>
</details>

**Comment by rmcgibbo** [comment]
Because of the way it's organized under pythonPackages, the attribute name that will be used in nixpkgs for this will be `python3Packages.packageurl-python`, or `python38Packages.packageurl-python` or something like that (e.g. the names that the bot shows in its comment), depending on the version of python. Given that, I think the convention would normally be to write the attribute name as just `packageurl`. And then the commit message should also be changed to `python3Packages.packageurl: init at 0.9.4`.

**Comment by rmcgibbo** [comment]
Here's an example of what the commit should look like adding yourself as a nixpkgs maintainer: https://github.com/NixOS/nixpkgs/commit/c1441787d290488ad470148b51d7fae9cbf38652

**Comment by armijnhemel** [comment]
> Here's an example of what the commit should look like adding yourself as a nixpkgs maintainer: [c144178](https://github.com/NixOS/nixpkgs/commit/c1441787d290488ad470148b51d7fae9cbf38652)

Urgh, I knew I forgot something. Will fix.

**Comment by armijnhemel** [comment]
> Because of the way it's organized under pythonPackages, the attribute name that will be used in nixpkgs for this will be `python3Packages.packageurl-python`, or `python38Packages.packageurl-python` or something like that (e.g. the names that the bot shows in its comment), depending on the version of python. Given that, I think the convention would normally be to write the attribute name as just `packageurl`, so that it's `python3Packages.packageurl` without the extra "-python". And then the commit message should also be changed to `python3Packages.packageurl: init at 0.9.4`. Also, perhaps you can reduce the description down to 1 sentence?

The package description *is* already one sentence (copied from the upstream project). I will remove the dot so it is more clear. The official package name seems to be "packageurl-python" so that's why I used it instead of simply "packageurl". I am agnostic about the name.

File: maintainers/maintainer-list.nix [hunk 1]
```diff
@@ -734,6 +734,12 @@
     githubId = 1296771;
     name = "Anders Riutta";
   };
+  armijnhemel = {
```
Comment (SuperSandro2000) [inline maintainers/maintainer-list.nix:737]: Please put this into a separate commit with the message ``maintainers: add armijnhemel``.

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]
Please follow the contributing guide when creating your commits.

File: pkgs/development/python-modules/packageurl-python/default.nix [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  pname = "packageurl-python";
+  version = "0.9.4";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "0mpvj8imsaqhrgfq1cxx16flc5201y78kqa7bh2i5zxsc29843mx";
+  };
+
+  meta = with lib; {
+    description = "A parser and builder for purl aka 'Package URLs' for Python 2 and 3.";
```
Comment (fabaff) [inline pkgs/development/python-modules/packageurl-python/default.nix:13]:
```suggestion
    description = "A parser and builder for purl aka 'Package URLs' for Python 2 and 3";
```

see https://nixos.org/manual/nixpkgs/unstable/#sec-standard-meta-attributes

Personally I would even go a step further and get rid of "purl" and the Python releases: 

```suggestion
    description = "Python parser and builder for package URLs";
```

File: pkgs/development/python-modules/packageurl-python/default.nix [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+buildPythonPackage rec {
+  pname = "packageurl-python";
+  version = "0.9.4";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "0mpvj8imsaqhrgfq1cxx16flc5201y78kqa7bh2i5zxsc29843mx";
+  };
+
```
Comment (fabaff) [inline pkgs/development/python-modules/packageurl-python/default.nix:11]: You are not able to run the tests:

```suggestion
  
  pythonImportsCheck = [ "packageurl" ];
```

**Review by fabaff — CHANGES_REQUESTED** [review]
The upstream repo contains [tests](https://github.com/package-url/packageurl-python/tree/master/tests). Please consider to use the GitHub as source.

File: pkgs/development/python-modules/packageurl-python/default.nix (hunk 2 above)
Comment (armijnhemel) [reply pkgs/development/python-modules/packageurl-python/default.nix:13]: This makes sense. Will change.

File: maintainers/maintainer-list.nix (hunk 1 above)
Comment (armijnhemel) [reply maintainers/maintainer-list.nix:737]: > Please follow the contributing guide when creating your commits.

In my days there wasn't even such a guide! :-P

Kidding aside, what is the best way forward? revert that particular commit?

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 113597` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
3 packages built:
</summary>
<ul>
<li>python37Packages.packageurl</li>
<li>python38Packages.packageurl</li>
<li>python39Packages.packageurl</li>
</ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

<details>
python37Packages.packageurl:
<details>

Zero tests run by pytest got detected: 'Ran 0 tests in 0.000s'
You have the following options to fix this:
  - Tell pytest(CheckHook) where to find the tests included in the package
  - Check if the GitHub Repo contains tests but they are not shipped with Pypi. If so please switch to `fetchFromGitHub`.
  - If the Packages does not contain any tests add `'doCheck = false;'` and a `pythonImportsCheck`.


Add a `checkPhase` for tests, or at least `pythonImportsCheck`.

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/python-include-tests.md
</details>
python38Packages.packageurl:
<details>

Zero tests run by pytest got detected: 'Ran 0 tests in 0.000s'
You have the following options to fix this:
  - Tell pytest(CheckHook) where to find the tests included in the package
  - Check if the GitHub Repo contains tests but they are not shipped with Pypi. If so please switch to `fetchFromGitHub`.
  - If the Packages does not contain any tests add `'doCheck = false;'` and a `pythonImportsCheck`.


Add a `checkPhase` for tests, or at least `pythonImportsCheck`.

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/python-include-tests.md
</details>
python39Packages.packageurl:
<details>

Zero tests run by pytest got detected: 'Ran 0 tests in 0.000s'
You have the following options to fix this:
  - Tell pytest(CheckHook) where to find the tests included in the package
  - Check if the GitHub Repo contains tests but they are not shipped with Pypi. If so please switch to `fetchFromGitHub`.
  - If the Packages does not contain any tests add `'doCheck = false;'` and a `pythonImportsCheck`.


Add a `checkPhase` for tests, or at least `pythonImportsCheck`.

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/python-include-tests.md
</details>
</details>

File: maintainers/maintainer-list.nix (hunk 1 above)
Comment (mweinelt) [reply maintainers/maintainer-list.nix:737]:
```
❯ git reset --soft HEAD~6
❯ git status
HEAD detached from FETCH_HEAD
Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
	modified:   maintainers/maintainer-list.nix
	new file:   pkgs/development/python-modules/packageurl-python/default.nix
	modified:   pkgs/top-level/python-packages.nix
❯ git restore --staged pkgs
❯ git status
HEAD detached from FETCH_HEAD
Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
	modified:   maintainers/maintainer-list.nix

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   pkgs/top-level/python-packages.nix

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	pkgs/development/python-modules/packageurl-python/
```

And then start commit things one by one again.

**Comment by armijnhemel** [comment]
well, that git reworking didn't work for me and I have no time to figure stuff out right now. I will close this one and will recreate it.

**Comment by armijnhemel** [comment]
resubmitted as https://github.com/NixOS/nixpkgs/pull/116325
